File: examples/basic_eval.rs
//! Basic example: compile and evaluate `2*x + 3` with a single value.

use std::collections::HashMap;

use mathlex::{BinaryOp, ExprKind, Expression};

use mathlex_eval::{EvalInput, NumericResult, compile, eval};

fn main() {
    // Build AST for: 2*x + 3
    let ast = ExprKind::Binary {
        op: BinaryOp::Add,
        left: Box::new(
            ExprKind::Binary {
                op: BinaryOp::Mul,
                left: Box::new(Expression::integer(2)),
                right: Box::new(Expression::variable("x")),
            }
            .into(),
        ),
        right: Box::new(Expression::integer(3)),
    }
    .into();

    // Compile with no constants
    let compiled = compile(&ast, &HashMap::new()).expect("compilation failed");
    println!("Arguments: {:?}", compiled.argument_names());
    println!("Complex: {}", compiled.is_complex());

    // Evaluate with x = 5.0
    let mut args = HashMap::new();
    args.insert("x", EvalInput::Scalar(5.0));
    let handle = eval(&compiled, args).expect("eval failed");
    let result = handle.scalar().expect("scalar eval failed");

    match result {
        NumericResult::Real(v) => println!("2*5 + 3 = {v}"),
        NumericResult::Complex(c) => println!("2*5 + 3 = {c}"),
    }
}

File: docs/swift-examples/BasicEval.swift
/// Basic evaluation example.
///
/// Demonstrates compiling and evaluating a simple expression.
///
/// Prerequisites:
///   cargo build --release --features ffi

import MathLexEval

func basicEvalExample() throws {
    // AST JSON for: 2*x + 3
    // (In practice, mathlex generates this from parsing "2*x + 3")
    let astJson = """
    {"Binary":{"op":"Add","left":{"Binary":{"op":"Mul","left":{"Integer":2},"right":{"Variable":"x"}}},"right":{"Integer":3}}}
    """

    // Compile with no constants
    let expr = try MathEvaluator.compile(json: astJson)
    print("Arguments: \(expr.argumentNames)")
    print("Complex: \(expr.isComplex)")

    // Evaluate with x = 5.0
    let result = try MathEvaluator.evaluate(expr, args: ["x": .scalar(5.0)])

    switch result {
    case .real(let v):
        print("2*5 + 3 = \(v)")
    case .complex(let re, let im):
        print("2*5 + 3 = \(re) + \(im)i")
    }
}
